#1636 Get rid of Lombok in langchain4j-core

kpavlov opened the pull request:
<!--
Thank you so much for your contribution!

Please fill in all the sections below.
Please open the PR as a draft initially. Once it is reviewed and approved, we will ask you to add documentation and examples.
Please note that PRs with breaking changes or without tests will be rejected.

Please note that PRs will be reviewed based on the priority of the issues they address.
We ask for your patience. We are doing our best to review your PR as quickly as possible.
Please refrain from pinging and asking when it will be reviewed. Thank you for understanding!
-->

## Issue
<!-- Please specify the ID of the issue this PR is addressing. For example: "Closes #1234" or "Fixes #1234" -->
Contributes to #1636 

## Change
<!-- Please describe the changes you made. -->
Get rid of Lombok in langchain4j-core: Run Delombok refactoring in IntelliJ IDEA to remove Lombok annotations and replace them with the equivalent Java code.

This pull request focuses on removing the Lombok dependency and replacing it with manually implemented builder patterns across several classes. Additionally, it includes some minor code improvements.

## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [x] There are no breaking changes
- [ ] I have added unit and integration tests for my change
- [ ] I have manually run all the unit tests in all modules, and they are all green
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

Commit:
#1636 get rid of Lombok in langchain4j-core

Commit:
#1636 get rid of Lombok in langchain4j-core

Commit:
#1636 get rid of Lombok in langchain4j-core

Commit:
#1636 get rid of Lombok in langchain4j-core

Commit:
#1636 get rid of Lombok in langchain4j-core

kpavlov committed:
Merge branch 'main' into delombok-core

Commit:
#1636 get rid of Lombok in langchain4j-core tests

glaforge commented on langchain4j-core/src/test/java/dev/langchain4j/agent/tool/ToolSpecificationsTest.java:
```diff
@@ -1,39 +1,26 @@
-import dev.langchain4j.model.chat.request.json.JsonObjectSchema;
-import dev.langchain4j.model.chat.request.json.JsonSchemaElement;
-import dev.langchain4j.model.chat.request.json.JsonStringSchema;
+import dev.langchain4j.model.chat.request.json.*;
```
Try to keep individual imports, and avoid using star imports.

kpavlov replied:
Sorry, missed that. Will add .editorconfig

kpavlov committed:
Merge branch 'main' into delombok-core

Commit:
Add imports for fully qualified references to types

https://docs.openrewrite.org/recipes/java/shortenfullyqualifiedtypereferences

Commit:
Add .editorconfig

Commit:
Optimize imports

Replace star imports with FQCN/FQMN

glaforge reviewed (dismissed):
Those changes look great to me!

kpavlov commented:
Looking forward for having Codacy or other quality enforcement tool to be integrated:)

Wadimz commented on langchain4j-core/src/main/java/dev/langchain4j/rag/DefaultRetrievalAugmentor.java:
```diff
@@ -264,12 +272,16 @@ private static void log(Query query, ContentRetriever retriever, List<Content> c
                 contents.size(), query.text(), retriever);
 
-        if (contents.size() > 0) {
+        if (!contents.isEmpty()) {
```
can we add here ` && log.isTraceEnabled()` to not build contentsString when logging was not requested?

Wadimz reviewed (approved):
Thanks for format and `*` import unwrapping!

kpavlov committed:
Merge branch 'main' into delombok-core

Commit:
Refactor `DefaultRetrievalAugmentor` logging and formatting

This commit updates the logging logic and re-formats the code in `DefaultRetrievalAugmentor.java`. It improves debug and trace message conditions and standardizes the indentation for better readability. Non-essential imports have also been removed.

Wadimz commented on langchain4j-core/src/main/java/dev/langchain4j/rag/DefaultRetrievalAugmentor.java:
```diff
@@ -227,62 +225,69 @@ private static List<Content> retrieve(ContentRetriever retriever, Query query) {
                     retriever.getClass().getName(),
                     contentsSting);
+            }
+        } else {
```
looks like else belongs to wrong if branch here. Should it be for inner if?

kpavlov replied:
[Fixed](https://github.com/langchain4j/langchain4j/pull/1987/files#diff-9a5519a26a4b6d2fd412c877de4459c2a119e793fece9385f558a93a7e0aee5aR273-R275)

Wadimz commented on langchain4j-core/src/main/java/dev/langchain4j/rag/DefaultRetrievalAugmentor.java:
```diff
@@ -227,62 +225,69 @@ private static List<Content> retrieve(ContentRetriever retriever, Query query) {
+                    originalQuery.text(), transformedQuery.text());
             }
-        } else {
+        } else if (log.isDebugEnabled()){
```
if we started updating logging, maybe we can wrap a whole content of this method with `if (log.isDebugEnabled())`. Looks all it does is logs with debug level

kpavlov replied:
Let's stop updating logging in this PR. Let's do it separately
Things are getting messy

kpavlov committed:
Merge branch 'main' into delombok-core

Commit:
Refactor trace logging in DefaultRetrievalAugmentor

Reorganized the conditional check for trace logging to exit early if trace logging is not enabled. This change reduces nesting and improves code readability.

Commit:
Update CONTRIBUTING.md with additional guidelines

Added steps to run `mvn dependency:analyze` to identify unnecessary dependencies and to execute all unit tests with `mvn clean test`. These changes ensure contributors check for redundant dependencies and verify their contributions against all current tests.

dliubarskyi committed:
Update langchain4j-core/src/main/java/dev/langchain4j/rag/DefaultRetrievalAugmentor.java
